File: src/thoth/thoth.c
/*
File Name:   Hermes.c
//
Compiler:    GCC
Author:      Artic42
Description: This is library helps create log file in UNIX like systems
Tested on:    Raspberry Pi OS and wsl2
*/

/****************************************
*	Include                             *
****************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "Artic42.h"
#include "thoth.h"

/****************************************
*   Private function prototype          *
****************************************/

void getMaskName (int mask);
void calculateString  (str Message, int mask);
void addLog2List (struct log* logPointer);
void removeLog2List (struct log* logPointer);
FILE* openLogFile (struct log* logPointer);
void calculateFirstLine (struct log* logPointer, dateTime today);
void closeLogFile (struct log* logPointer);

/****************************************
*   Private global variables            *
****************************************/

struct log *firstLog = NULL;
struct log *lastLog = NULL;
char messageBuffer [MAX_MSG_LENGTH+512];
char pathBuffer [MAX_PATH_LENGTH+MAX_NAME_LENGTH+50];
char maskBuffer [32];
char lineBuffer [MAX_FILE_LENGTH+50];

/****************************************
*   Private constants                   *
****************************************/

#define ENTRY_LIMIT 100

/****************************************
*   Code                                *
****************************************/

struct log* createLog (str name, str path, int mask, int entries)
{
    struct log *logPointer;

    logPointer = malloc (sizeof(struct log));

    strcpy(logPointer->name, name);
    strcpy(logPointer->directoryPath, path);
    logPointer->mask = mask;
    logPointer->entriesOnFile = 0;
    logPointer->entryLimit = entries;
    logPointer->nxtLog = NULL;
    logPointer->prvLog = NULL;
    logPointer->filePointer = openLogFile (logPointer);

    addLog2List (logPointer);

    return logPointer;
}

void closeLog (struct log *logPointer)
{
    closeLogFile (logPointer);
    removeLog2List (logPointer);
    free (logPointer);
}

void closeAllLogs (void)
{
    struct log *logPointer = firstLog;

    while (logPointer != NULL)
    {
        struct log *nxtPointer = logPointer->nxtLog;
        closeLog (logPointer);
        logPointer = nxtPointer;
    }
}

void addEntry (int mask, str message)
{
    struct log *logPointer = firstLog;

    calculateString (message, mask);

    while (logPointer != NULL)
    {
        if (mask & logPointer->mask)
        {
            addLine2File (messageBuffer, logPointer->filePointer);

            logPointer->entriesOnFile++;

            if (logPointer->entriesOnFile >= logPointer->entryLimit)
            {
                logPointer->entriesOnFile = 0;
                closeLogFile (logPointer);
                logPointer->filePointer = openLogFile (logPointer);
            }
        }
        logPointer = logPointer->nxtLog;
    }
}

void addMask (struct log* logPointer, int mask)
{
    logPointer->mask = logPointer->mask | mask;
}

void delMask (struct log* logPointer, int mask)
{
    logPointer->mask = logPointer->mask & (~mask);
}

void getMaskName (int mask)
{
    switch (mask)
    {
    case INFO:
        strcpy(maskBuffer, "INFO");
        break;
    case DEBUG:
        strcpy(maskBuffer, "DEBUG");
        break;
    case WARNING:
        strcpy(maskBuffer, "WARNING");
        break;
    case SAFETY_WARNING:
        strcpy(maskBuffer, "SAFETY_WARNING");
        break;
    case ALARM:
        strcpy(maskBuffer, "ALARM");
        break;
    case SAFETY_ALARM:
        strcpy(maskBuffer, "SAFETY_ALARM");
        break;
    case COMUNNICATION_SEND:
        strcpy(maskBuffer, "COMUNNICATION_SEND");
        break;
    case COMUNNICATION_REC:
        strcpy(maskBuffer, "COMUNNICATION_REC");
        break;
    case ERROR:
        strcpy(maskBuffer, "ERROR");
        break;
    case HERMES:
        strcpy(maskBuffer, "THOTH");
        break;
    default:
        break;
    }
}

void calculateString  (str Message, int mask)
{
    char partBuffer[MAX_MSG_LENGTH];
    dateTime today = getDateTime();

    getMaskName(mask);
    // YY/MM/DD HH:MM:SS - MASK - MESSAGE
    sprintf (partBuffer, "%04d/%02d/%02d ", today.year, today.month, today.day);
    strcpy (messageBuffer, partBuffer);

    sprintf (partBuffer, "%02d:%02d:%02d - ", today.hour, today.min, today.sec);
    strcat (messageBuffer, partBuffer);

    strcat (messageBuffer, maskBuffer);
    strcat (messageBuffer, " - ");
    strcat (messageBuffer, Message);
}

void addLog2List (struct log* logPointer)
{
    if (firstLog == NULL) { firstLog = logPointer; }
    logPointer->prvLog = lastLog;
    if (lastLog != NULL) { lastLog->nxtLog = logPointer; }
    lastLog = logPointer;
}

void removeLog2List (struct log* logPointer)
{
    if (logPointer->prvLog != NULL) {logPointer->prvLog->nxtLog = logPointer->nxtLog;}
    if (logPointer->nxtLog != NULL) {logPointer->nxtLog->prvLog = logPointer->prvLog;}
    if (logPointer->nxtLog == NULL) {lastLog = logPointer->prvLog;}
    if (logPointer->prvLog == NULL) {firstLog = logPointer->nxtLog;}
}

FILE* openLogFile (struct log* logPointer)
{
    FILE* FP;
    dateTime today = getDateTime();
    char partBuffer[MAX_MSG_LENGTH + 50];

    strcpy(pathBuffer, logPointer->directoryPath);
    strcat(pathBuffer, "/");
    strcat(pathBuffer, logPointer->name);

    sprintf(partBuffer, "_%04d%02d%02d_%02d%02d%02d.log", today.year, today.month, today.day, today.hour, today.min, today.sec);
    strcat(pathBuffer, partBuffer);

    createDirectory (logPointer->directoryPath);

    if (pathExists(pathBuffer) == BFALSE)
    {
        FP = openFile2Write (pathBuffer);
        calculateFirstLine (logPointer, today);
        addLine2File (lineBuffer, FP);
    }
    else
    {
        FP = openFile2Append (pathBuffer);
    }
    return FP;
}

void calculateFirstLine (struct log* logPointer, dateTime today)
{
    char dateString1 [32], dateString2 [32];
    sprintf(dateString1, "%04d/%02d/%02d ", today.year, today.month, today.day);
    sprintf (dateString2, "%02d:%02d:%02d", today.hour, today.min, today.sec);
    strcpy (lineBuffer, logPointer->name);
    strcat (lineBuffer, " - ");
    strcat (lineBuffer, dateString1);
    strcat (lineBuffer, dateString2);

}

void closeLogFile (struct log* logPointer)
{
    closeFile (logPointer->filePointer);
}

File: src/thoth/thoth.py
import date_time
import fileManagement

#Mask constants
INFO = 0x01
DEBUG = 0x02
WARNING = 0x04
SAFETY_WARNING = 0x08
ALARM = 0x10
SAFETY_ALARM = 0x20
COMMUNICATION_SEND = 0x40
COMMUNICATION_REC = 0x80
ERROR = 0x100
HERMES = 0x200

global listOfLogs
listOfLogs = []

def addEntry (mask, message):
    line = calculateString (message, mask)

    for log in listOfLogs:
        if (log.mask & mask) != 0x0:
            log.filePointer.writeLine (line)
            log.entriesOnFile = log.entriesOnFile + 1
            if log.entriesOnFile >= log.entryLimit:
                log.entriesOnFile = 0
                log.closeLogFile()
                log.filePointer = log.openLogFile()

def closeAllLogs ():
    listOfLogsCopy = listOfLogs.copy()
    for log in listOfLogsCopy:
        log.closeLog()

def getMaskName (mask):
    if mask == INFO:
        return "INFO"
    if mask == DEBUG:
        return "DEBUG"
    if mask == WARNING:
        return "WARNING"
    if mask == SAFETY_WARNING:
        return "SAFETY_WARNING"
    if mask == ALARM:
        return "ALARM"
    if mask == SAFETY_ALARM:
        return "SAFETY_ALARM"
    if mask == COMMUNICATION_SEND:
        return "COMMUNICATION_SEND"
    if mask == COMMUNICATION_REC:
        return "COMMUNICATION_REC"
    if mask == ERROR:
        return "ERROR"
    if mask == HERMES:
        return "HERMES"

def calculateString (Message, mask):
    date = date_time.createDate()
    date.now()
    result1 = date.createString()
    result2 = " - " + getMaskName (mask) + " - "
    return result1 + result2 + Message

class log:
#Public methods
    def __init__ (self, name, path, mask, entries):
        self.name = name
        self.directoryPath = path
        self.mask = mask
        self.entriesOnFile = 0
        self.entryLimit = entries
        self.filePointer = self.openLogFile ()
        
        listOfLogs.append (self)

    def closeLog (self):
        self.closeLogFile()
        listOfLogs.remove(self)
        del self

    def addMask (self, mask):
        self.mask = self.mask | mask

    def delMask (self, mask):
        self.mask = self.mask & (~mask)

#Private methods
    def openLogFile (self):
        path1 = self.directoryPath + "/" + self.name + "_"
        date = date_time.createDate()
        date.now()
        path2 = date.year_s + date.month_s + date.day_s + "_"
        path3 = date.hour_s + date.minute_s + date.second_s + ".log"
        path = path1 + path2 + path3

        fileManagement.createDir(self.directoryPath)
        file = fileManagement.fileIO()

        if fileManagement.checkExistsFile(path) == False:
            file.openFile2Write(path)
            creationDate = date_time.createDate()
            creationDate.now()
            file.writeLine(f"{self.name} - {creationDate.createString()}")
        else:
            file.openFile2Append(path)
        
        return file

    def closeLogFile (self):
        self.filePointer.closeFile()
